### PR TITLE
Added ocr data validation for new application

### PR DESCRIPTION
**Pull request by aliveni**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/BPS-398

### Change description ###
Added validation to check ocr data exists for new applications.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java [hunk 1]
```diff
@@ -147,6 +150,28 @@ public static void assertEnvelopeHasPdfs(InputEnvelope envelope, List<Pdf> pdfs)
...
+     *
+     * @param envelope to assert against
+     */
+    public static void assertEnvelopeContainsOcrDataForNewApplication(InputEnvelope envelope) {
+
+        if (Classification.NEW_APPLICATION.equals(envelope.classification)) {
+
+            boolean isValidScannableItems = envelope.scannableItems
+                .stream()
+                .anyMatch(item -> item.documentType.equals(InputDocumentType.SSCS1) && item.ocrData.isEmpty());
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java]: so `isValidScannableItems` means specific doc type and empty ocr data. isn't the requirement for ocr data to be present instead? if yes then giving such conditions as is now, the variable proposes to be `isNotValid...`

Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java]: No, ocr data shouldn't be empty. It is also missing jurisdiction condition. I'll push the changes soon

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java [hunk 2]
```diff
@@ -147,6 +151,29 @@ public static void assertEnvelopeHasPdfs(InputEnvelope envelope, List<Pdf> pdfs)
         }
     }
 
+    /**
+     * Assert envelope contains at least one scannable item
+     * with valid document type and ocr data for new application.
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java]: No mention of the fact that this is only checked for SSCS jurisdiction.

This function seems very specific the way it is right now.
Perhaps it could take parameters?

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/EnvelopeProcessorValidationTest.java [hunk 3]
```diff
@@ -76,6 +77,31 @@ public void should_throw_exception_when_zip_file_has_mismatching_pdf() throws Ex
             .hasMessageContaining("Missing PDFs: 1111001.pdf");
     }
 
+    @Test
+    public void should_throw_exception_when_ocr_data_is_missing_for_sscs_new_application() throws Exception {
```
Comment (lgonczar) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/EnvelopeProcessorValidationTest.java]: It would also be worth adding a test for BlobProcessorTask, checking if this method is used.

File: src/integrationTest/resources/zipcontents/missing_ocr_data/metadatajson.json [hunk 4]
```diff
@@ -0,0 +1,21 @@
+{
+  "po_box": "SSCSPO",
+  "jurisdiction": "SSCS",
+  "delivery_date": "2018-06-23T00:00:00.000Z",
+  "opening_date": "2018-06-24T00:00:00.000Z",
+  "zip_file_createddate": "2018-02-24T00:00:00.000Z",
+  "zip_file_name": "8_24-06-2018-00-00-00.zip",
+  "envelope_classification": "new_application",
+  "scannable_items": [
```
Comment (lgonczar) [inline src/integrationTest/resources/zipcontents/missing_ocr_data/metadatajson.json]: Maybe it would also be good to test a case when there's no SSCS1 document among scannable items (which can contain OCR data)?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java [hunk 5]
```diff
@@ -147,6 +151,29 @@ public static void assertEnvelopeHasPdfs(InputEnvelope envelope, List<Pdf> pdfs)
...
+     * Throws exception otherwise.
+     *
+     * @param envelope to assert against
+     */
+    public static void assertEnvelopeContainsOcrDataForNewApplication(InputEnvelope envelope) {
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java]: Currently, we've got a mix of static and non-static `assert...` methods. It would be best if all of them were the same. My preference would be to make them all non-static, as it's easier to test code with those. What do you think?

Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java]: I think non-static is better

File: src/integrationTest/resources/zipcontents/missing_ocr_data/metadatajson.json (hunk 4 above)
Comment (aliveni) [reply src/integrationTest/resources/zipcontents/missing_ocr_data/metadatajson.json]: 👍

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java [hunk 6]
```diff
@@ -147,6 +151,29 @@ public static void assertEnvelopeHasPdfs(InputEnvelope envelope, List<Pdf> pdfs)
...
+
+        if (envelope.jurisdiction.equalsIgnoreCase("SSCS")
+            && Classification.NEW_APPLICATION.equals(envelope.classification)) {
+
+            boolean isValidScannableItems = envelope.scannableItems
+                .stream()
+                .anyMatch(item -> item.documentType.equals(InputDocumentType.SSCS1) && !MapUtils.isEmpty(item.ocrData));
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java]: should probably check that **all** sscs1 documents have ocr data, not just at least one of them

Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java]: There was a long discussion about this at some point and as far as I remember we couldn't get any more specific than that we definitely require at least one sscs1 document with ocr data.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java [hunk 7]
```diff
@@ -236,13 +237,15 @@ private ZipFileProcessingResult processZipFileContent(
 
             InputEnvelope envelope = envelopeProcessor.parseEnvelope(result.getMetadata(), zipFilename);
 
+            EnvelopeProcessor.assertEnvelopeContainsOcrDataForNewApplication(envelope);
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java]: perhaps this could be called only for SSCS

Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java]: or `assertEnvelopeContainsOcrDataIfRequired` or sth like that

**Comment by rkondratowicz** [comment]
Should error notification be sent? if so, an exception needs to be registered in `ErrorMapping` (and test added ;))

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/EnvelopeProcessorValidationTest.java (hunk 3 above)
Comment (aliveni) [reply src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/EnvelopeProcessorValidationTest.java]: We don't have a unit test for BlobProcessorTask. I think it's better to do it as part of another ticket

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java (hunk 7 above)
Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java]: I couldn't think of any other better name, so I renamed it as `assertEnvelopeContainsOcrDataIfRequired`

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java (hunk 2 above)
Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java]: Updated javadoc to include the conditions for jurisdiction and document_type

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java [hunk 8]
```diff
@@ -117,7 +121,7 @@ public Envelope getEnvelopeByFileAndContainer(String container, String zipFileNa
      * @param envelope to assert against
      * @param pdfs     to assert against
      */
-    public static void assertEnvelopeHasPdfs(InputEnvelope envelope, List<Pdf> pdfs) {
+    public void assertEnvelopeHasPdfs(InputEnvelope envelope, List<Pdf> pdfs) {
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java]: unrelated, unnecessary change imho

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java [hunk 9]
```diff
@@ -147,6 +151,32 @@ public static void assertEnvelopeHasPdfs(InputEnvelope envelope, List<Pdf> pdfs)
...
+     */
+    public void assertEnvelopeContainsOcrDataIfRequired(InputEnvelope envelope) {
+
+        if (envelope.jurisdiction.equalsIgnoreCase("SSCS")
+            && Classification.NEW_APPLICATION.equals(envelope.classification)) {
+
+            boolean ocrDataExists = envelope.scannableItems
+                .stream()
+                .filter(item -> item.documentType.equals(InputDocumentType.SSCS1))
+                .noneMatch(item -> MapUtils.isEmpty(item.ocrData));
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java]: that doesn't mean ocr data exists, when there are no SSCS1 documents (ocr can be in a document of different type)

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java (hunk 8 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java]: more validation coming soon, let's leave it as is, I'll move validation to a separate class on next PR once this is merged.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java (hunk 7 above)
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java]: I meant 'named', though :)
